### PR TITLE
MINIFICPP-1429 clean up pre-c++17 workarounds

### PR DESCRIPTION
**Pull request by szaszm**
due to previously missing language and stdlib features

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: encrypt-config/ArgParser.h [hunk 1]
```diff
@@ -48,20 +48,20 @@ class Arguments {
...
+  static std::optional<Argument> getArg(const std::string& key);
+  static std::optional<Flag> getFlag(const std::string& flag);
 
  public:
   static Arguments parse(int argc, char* argv[]);
 
   static std::string getHelp();
 
-  utils::optional<std::string> get(const std::string& key) const;
+  [[nodiscard]] std::optional<std::string> get(const std::string& key) const;
```
Comment (szaszm) [inline encrypt-config/ArgParser.h:59]: I added a bunch of these `[[nodiscard]]` attributes where my IDE showed clang-tidy suggestions to do so.

**Review by fgerlits — APPROVED** [review]
LGTM

File: libminifi/include/utils/StringUtils.h [hunk 2]
```diff
@@ -17,24 +17,23 @@
...
-#include <sstream>
-#include <vector>
-#include <map>
-#include <type_traits>
 #include "utils/FailurePolicy.h"
-#include "utils/GeneralUtils.h"
 #include "utils/gsl.h"
-#include "utils/OptionalUtils.h"
 
 #if defined(WIN32) || (__cplusplus >= 201103L && (!defined(__GLIBCXX__) || (__cplusplus >= 201402L) ||  (defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE > 4)))
```
Comment (martinzink) [inline libminifi/include/utils/StringUtils.h:38]: I think this is always true now, and it could be simplified.

**Review by martinzink — COMMENTED** [review]
LGTM one minor comment

File: libminifi/include/utils/StringUtils.h (hunk 2 above)
Comment (szaszm) [reply libminifi/include/utils/StringUtils.h:38]: The `HAVE_REGEX_CPP` macro doesn't seem to be used anywhere, so I'm removing the whole block. I'll leave the removal of utils::Regex to a followup: https://issues.apache.org/jira/browse/MINIFICPP-1508

**Review by martinzink — APPROVED** [review]

File: extensions/systemd/ConsumeJournald.cpp [hunk 3]
```diff
@@ -23,6 +23,7 @@
 #include "date/date.h"
 #include "spdlog/spdlog.h"  // TODO(szaszm): make fmt directly available
 #include "utils/GeneralUtils.h"
+#include "utils/OptionalUtils.h"
```
Comment (lordgamez) [inline extensions/systemd/ConsumeJournald.cpp:26]: This shouldn't be needed.

Comment (szaszm) [reply extensions/systemd/ConsumeJournald.cpp:26]: It's needed for utils::map, utils::flatMap and utils::orElse. Those are still our own and all of them are used in this file.

Comment (lordgamez) [reply extensions/systemd/ConsumeJournald.cpp:26]: I see, thanks!

**Review by lordgamez — APPROVED** [review]

File: libminifi/test/unit/ProcessSessionTests.cpp [hunk 4]
```diff
@@ -31,25 +32,16 @@ REGISTER_RESOURCE(DummyProcessor, "A processor that does nothing.");
...
-  std::shared_ptr<core::Processor> dummy_processor_;
-  std::shared_ptr<core::ProcessContext> context_;
-  std::unique_ptr<core::ProcessSession> process_session_;
+  std::shared_ptr<TestPlan> test_plan_ = test_controller_.createPlan();
+  std::shared_ptr<core::Processor> dummy_processor_ = test_plan_->addProcessor("DummyProcessor", "dummyProcessor");
+  std::shared_ptr<core::ProcessContext> context_ = [this] { test_plan_->runNextProcessor(); return test_plan_->getCurrentContext(); }();
+  std::unique_ptr<core::ProcessSession> process_session_ = std::make_unique<core::ProcessSession>(context_);
 };
 
-Fixture::Fixture() {
```
Comment (arpadboda) [inline libminifi/test/unit/ProcessSessionTests.cpp:45]: What's the motivation behind deleting this?

**Review by arpadboda — APPROVED** [review]
Nice to see such much code being deleted!

On question, otherwise LGTM

File: libminifi/test/unit/ProcessSessionTests.cpp (hunk 4 above)
Comment (szaszm) [reply libminifi/test/unit/ProcessSessionTests.cpp:45]: Just wanted to simplify the code by moving the initialization code to default member initializers. I can revert if you have any objections.

Comment (arpadboda) [reply libminifi/test/unit/ProcessSessionTests.cpp:45]: No, just didn't find the place in the diff where it was actually moved to. :) Got it, thanks!

**Pull request closed without merging**